# Song analysis

In [34]:
import pandas as pd
import numpy as np

## Data cleaning

1. **spotify-top-200**

In [35]:
top_200_df = pd.read_csv("data/spotify-top-200-dataset.csv", sep = ";")
top_200_df.head()

,track_id,track_name,track_popularity,track_number,album_id,album_name,album_img,album_type,album_label,album_track_number,...,speechiness,acousticness,instrumentalness,liveness,valence,tempo,duration,pivot,streams,track_index
0,5aAx2yezTd8zXrkmtKl66Z,Starboy,0,1,09fggMHib4YkOtwQNXEBII,Starboy,https://i.scdn.co/image/ab67616d0000b2730c8599...,album,Universal Music Group,1,...,0.2820,0.165,0.000003,0.134,0.535,186.054,230453,0,25734078,1
1,5aAx2yezTd8zXrkmtKl66Z,Starboy,0,1,09fggMHib4YkOtwQNXEBII,Starboy,https://i.scdn.co/image/ab67616d0000b2730c8599...,album,Universal Music Group,1,...,0.2820,0.165,0.000003,0.134,0.535,186.054,230453,1,25734078,1
2,7BKLCZ1jbUBVqRi2FVlTVw,Closer,84,1,0rSLgV8p5FzfnqlEk4GzxE,Closer,https://i.scdn.co/image/ab67616d0000b273495ce6...,single,Disruptor Records/Columbia,1,...,0.0338,0.414,0.000000,0.111,0.661,95.010,244960,0,23519705,2
3,7BKLCZ1jbUBVqRi2FVlTVw,Closer,84,1,0rSLgV8p5FzfnqlEk4GzxE,Closer,https://i.scdn.co/image/ab67616d0000b273495ce6...,single,Disruptor Records/Columbia,1,...,0.0338,0.414,0.000000,0.111,0.661,95.010,244960,1,23519705,2
4,5knuzwU65gJK7IF5yJsuaW,Rockabye (feat. Sean Paul & Anne-Marie),75,1,3meZFplbMmji648oWUNEfQ,Rockabye (feat. Sean Paul & Anne-Marie),https://i.scdn.co/image/ab67616d0000b2731431c3...,single,Atlantic Records UK,1,...,0.0523,0.406,0.000000,0.180,0.742,101.965,251088,0,21216399,3


In [36]:
# check data types
top_200_df.info()

<class 'pandas.DataFrame'>
RangeIndex: 74660 entries, 0 to 74659
Data columns (total 40 columns):
 #   Column              Non-Null Count  Dtype  
---  ------              --------------  -----  
 0   track_id            74660 non-null  str    
 1   track_name          74655 non-null  str    
 2   track_popularity    74660 non-null  int64  
 3   track_number        74660 non-null  int64  
 4   album_id            74660 non-null  str    
 5   album_name          74655 non-null  str    
 6   album_img           74655 non-null  str    
 7   album_type          74660 non-null  str    
 8   album_label         74655 non-null  str    
 9   album_track_number  74660 non-null  int64  
 10  album_popularity    74660 non-null  int64  
 11  artist_num          74660 non-null  int64  
 12  artist_names        74655 non-null  str    
 13  artist_id           74660 non-null  str    
 14  artist_name         74660 non-null  str    
 15  artist_img          74535 non-null  str    
 16  artist_follower

In [37]:
# show summary
top_200_df.describe()

,track_popularity,track_number,album_track_number,album_popularity,artist_num,artist_followers,artist_popularity,rank,danceability,energy,...,speechiness,acousticness,instrumentalness,liveness,valence,tempo,duration,pivot,streams,track_index
count,74660.000000,74660.000000,74660.000000,74660.000000,74660.000000,7.466000e+04,74660.000000,74660.000000,74660.000000,74660.000000,...,74660.000000,74660.000000,74660.000000,74660.000000,74660.000000,74660.000000,74660.000000,74660.000000,7.466000e+04,74660.000000
mean,57.851661,3.843892,3.843892,52.386244,2.233003,1.992528e+07,79.868711,99.574819,0.707997,0.651976,...,0.119321,0.217359,0.005648,0.167878,0.520031,120.762263,212258.769796,0.407983,8.336837e+06,21867.919314
std,29.514677,4.390576,4.390576,27.851247,1.320795,2.264422e+07,9.941376,57.293789,0.127662,0.152462,...,0.100752,0.219920,0.044550,0.122673,0.218686,29.824768,44088.043199,0.491463,5.698289e+06,12608.146829
min,0.000000,1.000000,1.000000,0.000000,1.000000,4.000000e+00,0.000000,1.000000,0.150000,0.027900,...,0.023200,0.000025,0.000000,0.019700,0.032000,46.718000,30133.000000,0.000000,2.525159e+06,1.000000
25%,57.000000,1.000000,1.000000,47.000000,1.000000,3.639519e+06,75.000000,50.000000,0.637000,0.552000,...,0.047600,0.048400,0.000000,0.093600,0.352000,96.172000,184560.000000,0.000000,4.945827e+06,10951.750000
50%,71.000000,1.000000,1.000000,62.000000,2.000000,1.065244e+07,81.000000,99.000000,0.727000,0.669000,...,0.077400,0.145000,0.000000,0.123000,0.518000,115.017500,207232.500000,0.000000,6.333818e+06,21846.000000
75%,77.000000,5.000000,5.000000,72.000000,3.000000,2.993942e+07,87.000000,149.000000,0.796000,0.769000,...,0.155000,0.310000,0.000012,0.196000,0.690000,140.060000,231906.000000,1.000000,9.642750e+06,32528.000000
max,92.000000,76.000000,76.000000,86.000000,11.000000,1.040598e+08,100.000000,200.000000,0.980000,0.978000,...,0.966000,0.994000,0.953000,0.962000,0.982000,212.117000,577199.000000,1.000000,8.076404e+07,44200.000000


In [38]:
# check # of columns and rows
top_200_df.shape

(74660, 40)

## Filtering

There are 40 columns and 74660 rows in top_200_df. 
Since there are more than 20 columns and 2000 rows in top_200_df, we will filter and keep columns and rows which we need for this project.


In [39]:
# keep track_id, track_name, artist_name, release_date, week, and rank
top_200_df = top_200_df[[
    "track_id",
    "track_name",
    "artist_name",
    "release_date",
    "week",
    "rank"
]]

# standarize track_name and artist_name
top_200_df["track_name"] = top_200_df["track_name"].str.lower().str.strip()
top_200_df["artist_name"] = top_200_df["artist_name"].str.lower().str.strip()

# convert date 
top_200_df["release_date"] = pd.to_datetime(top_200_df["release_date"], errors = "coerce")
top_200_df["week"] = pd.to_datetime(top_200_df["week"], errors = "coerce")

# fliter 2020
top_200_df = top_200_df[top_200_df["week"].dt.year == 2020]

# remove duplicate
top_200_df = top_200_df.drop_duplicates(subset=["track_id"])

# remove missing track_id
top_200_df = top_200_df.dropna(subset=["track_id"])

# filtered data
top_200_df

,track_id,track_name,artist_name,release_date,week,rank
53071,1rgnBhdG2JDFTbYkYRZAku,dance monkey,tones and i,2019-05-10,2020-01-03,1
53072,696DnlkuDOXcMAnKlTgXXK,roxanne,arizona zervas,2019-10-10,2020-01-03,2
53073,7k4t7uLgtOxPwTpFmtJNTY,tusa,karol g,2019-11-07,2020-01-03,3
53075,2b8fOow8UzyDFAE27YhOZM,memories,maroon 5,2019-09-20,2020-01-03,4
53076,4TnjEaWOeW0eKTKIEvJyCa,falling,trevor daniel,2018-10-05,2020-01-03,5
...,...,...,...,...,...,...
69968,2mvabkN1i2gLnGAPUVdwek,christmas tree farm,taylor swift,2019-12-06,2020-12-25,181
69969,02FaKXXL7KUtRc7K0k54tL,cozy little christmas,katy perry,2018-11-15,2020-12-25,182
69973,15coTBAzEN1bOeipoNDZAR,"merry christmas, happy holidays",*nsync,1998-10-23,2020-12-25,184
69982,3QIoEi8Enr9uHffwInGIsC,please come home for christmas - 2013 remaster,eagles,NaT,2020-12-25,192


2. **spotify-tracks**


In [40]:
# read dataset
spotify_track_df = pd.read_csv("data/dataset.csv")
spotify_track_df.head()

,Unnamed: 0,track_id,artists,album_name,track_name,popularity,duration_ms,explicit,danceability,energy,...,loudness,mode,speechiness,acousticness,instrumentalness,liveness,valence,tempo,time_signature,track_genre
0,0,5SuOikwiRyPMVoIQDJUgSV,Gen Hoshino,Comedy,Comedy,73,230666,False,0.676,0.4610,...,-6.746,0,0.1430,0.0322,0.000001,0.3580,0.715,87.917,4,acoustic
1,1,4qPNDBW1i3p13qLCt0Ki3A,Ben Woodward,Ghost (Acoustic),Ghost - Acoustic,55,149610,False,0.420,0.1660,...,-17.235,1,0.0763,0.9240,0.000006,0.1010,0.267,77.489,4,acoustic
2,2,1iJBSr7s7jYXzM8EGcbK5b,Ingrid Michaelson;ZAYN,To Begin Again,To Begin Again,57,210826,False,0.438,0.3590,...,-9.734,1,0.0557,0.2100,0.000000,0.1170,0.120,76.332,4,acoustic
3,3,6lfxq3CG4xtTiEg7opyCyx,Kina Grannis,Crazy Rich Asians (Original Motion Picture Sou...,Can't Help Falling In Love,71,201933,False,0.266,0.0596,...,-18.515,1,0.0363,0.9050,0.000071,0.1320,0.143,181.740,3,acoustic
4,4,5vjLSffimiIP26QG5WcN2K,Chord Overstreet,Hold On,Hold On,82,198853,False,0.618,0.4430,...,-9.681,1,0.0526,0.4690,0.000000,0.0829,0.167,119.949,4,acoustic


In [41]:
# show summary
spotify_track_df.describe()

,Unnamed: 0,popularity,duration_ms,danceability,energy,key,loudness,mode,speechiness,acousticness,instrumentalness,liveness,valence,tempo,time_signature
count,114000.000000,114000.000000,1.140000e+05,114000.000000,114000.000000,114000.000000,114000.000000,114000.000000,114000.000000,114000.000000,114000.000000,114000.000000,114000.000000,114000.000000,114000.000000
mean,56999.500000,33.238535,2.280292e+05,0.566800,0.641383,5.309140,-8.258960,0.637553,0.084652,0.314910,0.156050,0.213553,0.474068,122.147837,3.904035
std,32909.109681,22.305078,1.072977e+05,0.173542,0.251529,3.559987,5.029337,0.480709,0.105732,0.332523,0.309555,0.190378,0.259261,29.978197,0.432621
min,0.000000,0.000000,0.000000e+00,0.000000,0.000000,0.000000,-49.531000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000
25%,28499.750000,17.000000,1.740660e+05,0.456000,0.472000,2.000000,-10.013000,0.000000,0.035900,0.016900,0.000000,0.098000,0.260000,99.218750,4.000000
50%,56999.500000,35.000000,2.129060e+05,0.580000,0.685000,5.000000,-7.004000,1.000000,0.048900,0.169000,0.000042,0.132000,0.464000,122.017000,4.000000
75%,85499.250000,50.000000,2.615060e+05,0.695000,0.854000,8.000000,-5.003000,1.000000,0.084500,0.598000,0.049000,0.273000,0.683000,140.071000,4.000000
max,113999.000000,100.000000,5.237295e+06,0.985000,1.000000,11.000000,4.532000,1.000000,0.965000,0.996000,1.000000,1.000000,0.995000,243.372000,5.000000


In [42]:
# check data type
spotify_track_df.info()

<class 'pandas.DataFrame'>
RangeIndex: 114000 entries, 0 to 113999
Data columns (total 21 columns):
 #   Column            Non-Null Count   Dtype  
---  ------            --------------   -----  
 0   Unnamed: 0        114000 non-null  int64  
 1   track_id          114000 non-null  str    
 2   artists           113999 non-null  str    
 3   album_name        113999 non-null  str    
 4   track_name        113999 non-null  str    
 5   popularity        114000 non-null  int64  
 6   duration_ms       114000 non-null  int64  
 7   explicit          114000 non-null  bool   
 8   danceability      114000 non-null  float64
 9   energy            114000 non-null  float64
 10  key               114000 non-null  int64  
 11  loudness          114000 non-null  float64
 12  mode              114000 non-null  int64  
 13  speechiness       114000 non-null  float64
 14  acousticness      114000 non-null  float64
 15  instrumentalness  114000 non-null  float64
 16  liveness          114000 non-nu

In [43]:
spotify_track_df.shape

(114000, 21)

## filtering

There are 21 columns and 114000 rows in spotify_track_df.
Since there are more than 20 columns and 2000 rows in the spotify_track_df, we filter and keep columns and rows which we need for this project.

In [44]:
# keep track_id, track_name, artists, and track_genre
spotify_track_df = spotify_track_df[[
    "track_id",
    "track_name",
    "artists",
    "track_genre"
]]

# standarize track_name and artist_name
spotify_track_df["track_name"] = spotify_track_df["track_name"].str.lower().str.strip()
spotify_track_df["artists"] = spotify_track_df["artists"].str.lower().str.strip()
spotify_track_df["track_genre"] = spotify_track_df["track_genre"].str.lower().str.strip()

# remove duplicate
spotify_track_df = spotify_track_df.drop_duplicates(subset=["track_id"])

# remove missing track_id
spotify_track_df = spotify_track_df.dropna(subset=["track_id"])

# before this filtering there are still 90k rows
# filter tracks appearing in top_200_df
top_200_ids = set(top_200_df["track_id"])
spotify_track_df = spotify_track_df[spotify_track_df["track_id"].isin(top_200_ids)]

# filtered data
spotify_track_df

,track_id,track_name,artists,track_genre
20,3S0OXQeoh0w6AY8WQVckRW,i'm yours,jason mraz,acoustic
2000,5E30LdtzQTGqRvNd7l6kG5,daddy issues,the neighbourhood,alt-rock
2003,2QjOHCTQ1Jl3zawyYOpxh6,sweater weather,the neighbourhood,alt-rock
2151,7JDWhC422Gtk1Bq0mL3OTC,feel something,bea miller,alt-rock
3006,60a0Rd6pjrkxjPbaKzXjfq,in the end,linkin park,alternative
...,...,...,...,...
98250,0Ie5uiv54KgCr7P4sYDTHl,baby it's cold outside (duet with michael bublé),idina menzel;michael bublé,show-tunes
99017,3dYD57lRAUcMHufyqn9GcI,take me to church,hozier,singer-songwriter
103007,3U4isOIWM3VvDubwSI3y7a,all of me,john legend,soul
103157,1Cv1YLb4q0RzL6pybtaMLo,sunday best,surfaces,soul


3. **tiktok-trend-track**

In [45]:
# read tiktok dataset
tiktok_df = pd.read_csv("data/tiktok.csv")
tiktok_df.head()

,track_id,track_name,artist_id,artist_name,album_id,duration,release_date,popularity,danceability,energy,...,speechiness,acousticness,instrumentalness,liveness,valence,tempo,playlist_id,playlist_name,duration_mins,genre
0,6kVuF2PYLuvl9T85XjNbaO,Lay It Down Gmix - Main,1Xfmvd48oOhEWkscWyEbh9,Lloyd,43C6GVlhXG4KfZuEbxty3r,302186,2011-01-01,28,0.597,0.800,...,0.3120,0.0461,0.0,0.1800,0.565,155.932,6kVuF2PYLuvl9T85XjNbaO,6kVuF2PYLuvl9T85XjNbaO,5.036433,TIKTOK DANCE
1,1RGIjMFMgJxkZHMDXVYzOJ,Bartender (feat. Akon),3aQeKQSyrW4qWr35idm0cy,T-Pain,6CrSEKCF4TYrbSIitegb3h,238800,2007-06-05,75,0.832,0.391,...,0.0628,0.0564,0.0,0.2240,0.436,104.961,1RGIjMFMgJxkZHMDXVYzOJ,1RGIjMFMgJxkZHMDXVYzOJ,3.980000,TIKTOK DANCE
2,1RGIjMFMgJxkZHMDXVYzOJ,Bartender (feat. Akon),3aQeKQSyrW4qWr35idm0cy,T-Pain,6CrSEKCF4TYrbSIitegb3h,238800,2007-06-05,75,0.832,0.391,...,0.0628,0.0564,0.0,0.2240,0.436,104.961,1RGIjMFMgJxkZHMDXVYzOJ,1RGIjMFMgJxkZHMDXVYzOJ,3.980000,TIKTOK DANCE
3,1dIWPXMX4kRHj6Dt2DStUQ,Chosen (feat. Ty Dolla $ign),4qXC0i02bSFstECuXP2ZpL,Blxst,7AwrgenNcTAJlJF3pKL0Qr,161684,2020-12-04,76,0.571,0.767,...,0.2870,0.3360,0.0,0.0809,0.605,93.421,1dIWPXMX4kRHj6Dt2DStUQ,1dIWPXMX4kRHj6Dt2DStUQ,2.694733,TIKTOK DANCE
4,4QVS8YCpK71R4FsxSMCjhP,Tie Me Down (with Elley Duhé),2ZRQcIgzPCVaT9XKhXZIzh,Gryffin,69t8rpgBN1ov5kCU6LDMuR,218295,2018-08-03,72,0.548,0.839,...,0.0644,0.1350,0.0,0.1020,0.314,98.932,4QVS8YCpK71R4FsxSMCjhP,4QVS8YCpK71R4FsxSMCjhP,3.638250,TIKTOK DANCE


In [46]:
# show summary 
tiktok_df.describe()

,duration,popularity,danceability,energy,key,loudness,mode,speechiness,acousticness,instrumentalness,liveness,valence,tempo,duration_mins
count,6746.000000,6746.000000,6746.000000,6746.000000,6746.000000,6746.000000,6746.000000,6746.000000,6746.000000,6746.000000,6746.000000,6746.000000,6746.000000,6746.000000
mean,194428.739253,57.649422,0.740926,0.624474,5.292914,-7.005657,0.581382,0.140059,0.219776,0.033784,0.177595,0.547327,120.525980,3.240479
std,59208.593470,24.616537,0.137951,0.174433,3.746116,2.846230,0.493369,0.128355,0.236456,0.138327,0.135816,0.230005,25.597213,0.986810
min,43426.000000,0.000000,0.150000,0.023700,0.000000,-26.888000,0.000000,0.023200,0.000010,0.000000,0.019300,0.033100,54.367000,0.723767
25%,155866.000000,44.000000,0.663000,0.504000,1.000000,-8.534000,0.000000,0.048600,0.031700,0.000000,0.092700,0.372000,100.053000,2.597767
50%,186980.000000,64.000000,0.757000,0.621000,6.000000,-6.609000,1.000000,0.084200,0.130000,0.000000,0.122000,0.543000,120.986000,3.116333
75%,224284.000000,76.000000,0.843000,0.752000,9.000000,-5.018000,1.000000,0.203000,0.321000,0.000215,0.224000,0.735000,135.001750,3.738067
max,716206.000000,100.000000,0.990000,0.999000,11.000000,1.077000,1.000000,0.911000,0.991000,0.962000,0.952000,0.998000,216.053000,11.936767


In [47]:
# check type
tiktok_df.info()

<class 'pandas.DataFrame'>
RangeIndex: 6746 entries, 0 to 6745
Data columns (total 23 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   track_id          6746 non-null   str    
 1   track_name        6746 non-null   str    
 2   artist_id         6746 non-null   str    
 3   artist_name       6746 non-null   str    
 4   album_id          6746 non-null   str    
 5   duration          6746 non-null   int64  
 6   release_date      6746 non-null   str    
 7   popularity        6746 non-null   int64  
 8   danceability      6746 non-null   float64
 9   energy            6746 non-null   float64
 10  key               6746 non-null   int64  
 11  loudness          6746 non-null   float64
 12  mode              6746 non-null   int64  
 13  speechiness       6746 non-null   float64
 14  acousticness      6746 non-null   float64
 15  instrumentalness  6746 non-null   float64
 16  liveness          6746 non-null   float64
 17  valenc

In [48]:
# check # of columns and rows
tiktok_df.shape

(6746, 23)

## Filtering 

There are 23 columns and 6746 rows in tiktok_df.
Since there are more than 20 columns and 2000 rows in the tiktok_df, we filter and keep columns and rows which we need for this project.

In [49]:
# keep track_id, track_name, artist_name, release_date, and popularity
tiktok_df = tiktok_df[[
    "track_id",
    "track_name",
    "artist_name",
    "release_date",
    "popularity"
]]

# standarize track_name and artist_name
tiktok_df["track_name"] = tiktok_df["track_name"].str.lower().str.strip()
tiktok_df["artist_name"] = tiktok_df["artist_name"].str.lower().str.strip()

# convert date 
tiktok_df["release_date"] = pd.to_datetime(tiktok_df["release_date"], errors = "coerce")

# since we have a lot of data, filter release_data in 2020
# filter the song data released in 2020
tiktok_df = tiktok_df[tiktok_df["release_date"].dt.year == 2020]

# remove duplicate
tiktok_df = tiktok_df.drop_duplicates(subset=["track_id"])

# filtered data
tiktok_df

,track_id,track_name,artist_name,release_date,popularity
3,1dIWPXMX4kRHj6Dt2DStUQ,chosen (feat. ty dolla $ign),blxst,2020-12-04,76
29,5mVeslUjgyVDBUB0bPJr6R,close up,prophet the artist,2020-11-11,50
31,3FAJ6O0NOHQV8Mc5Ri6ENp,heartbreak anniversary,giveon,2020-03-27,92
46,5GkQIP5mWPi4KZLLXeuFTT,motive (with doja cat),ariana grande,2020-10-30,82
79,0ZquDRUoxvC4QXHlBfRePZ,alo michael - rico rico rico rico,los dioses del ritmo,2020-05-30,35
...,...,...,...,...,...
6691,00iopF9xCuEX3XFxMUyO2M,m to the b,hiko,2020-09-03,43
6721,14tNUeer5gUbUYvhYYTclS,reveal,the boyz,2020-02-10,61
6724,4uqh9bualXNHXXwO2wPorc,daisies,katy perry,2020-05-15,71
6731,4kIpBfvK44bxqX7zo8K1oP,시작,gaho,2020-02-01,67


In [50]:
# just check if we have enough data 

top_200_ids = set(top_200_df["track_id"])
spotify_track_ids = set(spotify_track_df["track_id"])
tiktok_ids = set(tiktok_df["track_id"])

In [51]:
# number of same track_id between top_200_df and spotify_track_df
a = len(top_200_ids.intersection(spotify_track_ids))
a

232

In [52]:
# number of same track_id between top_200_df and tiktok_df
b = len(top_200_ids.intersection(tiktok_ids))
b

162

In [53]:
# number of same track_id between spotify_track_df and spotify_track_df
c = len(top_200_ids.intersection(spotify_track_ids).intersection(tiktok_ids))
c

45

In [54]:
# save cleaned data

top_200_df.to_csv("clean_spotify_top_200.csv", index = False)
spotify_track_df.to_csv("clean_spotify_track.csv", index = False)
tiktok_df.to_csv("clean_tiktok.csv", index = False)